### PR TITLE
fix(DetachEventBehavior): allow batch detach on elements.move

### LINKED ISSUE #1202: Boundary Events not detached when moved out of host in a batch move
Screen Recording 2019-09-25 at 11 22 08

__Describe the Bug__

The Boundary Events are not converted to Intermediate Events when moved to canvas in a batch operation.

__Steps to Reproduce__

1. Select multiple Boundary Events
2. Move them to canvas.

__Expected Behavior__

Such operation should either be disallowed or the events should be converted to Intermediate Events.

__Environment__

 - Browser: Chrome
 - OS: MacOS
 - Library version: 5.0.4

### PR DESCRIPTION
**Pull request by oguzeroglu**
Closes #1202

Kapture 2019-09-30 at 15 25 12

### CODE REVIEW

File: test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js [hunk 1]
```diff
@@ -61,6 +60,58 @@ describe('features/modeling/behavior - detach events', function() {
...
+      elements.push(subProcess);
+
+      // when
+      modeling.moveElements(elements, { x: 0, y: 300 }, root);
+
+      // then
+      eventIds.map(function(eventId) {
+        return elementRegistry.get(eventId);
+      }).forEach(function(element) {
+        console.log(element);
```
Comment (barmac) [inline test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js]: Let's remove that :)

File: test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js [hunk 2]
```diff
@@ -61,6 +60,58 @@ describe('features/modeling/behavior - detach events', function() {
...
+          }),
+          subProcess = elementRegistry.get('SubProcess_1'),
+          root = canvas.getRootElement();
+      elements.push(subProcess);
+
+      // when
+      modeling.moveElements(elements, { x: 0, y: 300 }, root);
+
+      // then
+      eventIds.map(function(eventId) {
```
Comment (barmac) [inline test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js]: Why don't we use `elements` here?

File: test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js [hunk 3]
```diff
@@ -61,6 +60,58 @@ describe('features/modeling/behavior - detach events', function() {
...
+    it('should not execute when moved with host', inject(function(canvas, elementRegistry, modeling) {
+
+      // given
+      var eventIds = [ 'BoundaryEventWithLabel', 'BoundaryEvent', 'BoundaryTimerEvent' ],
+          elements = eventIds.map(function(eventId) {
+            return elementRegistry.get(eventId);
+          }),
+          subProcess = elementRegistry.get('SubProcess_1'),
+          root = canvas.getRootElement();
+      elements.push(subProcess);
```
Comment (barmac) [inline test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js]:
```suggestion
      var movedElements = elements.concat(subProcess);
```

File: test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js [hunk 4]
```diff
@@ -61,6 +60,58 @@ describe('features/modeling/behavior - detach events', function() {
...
+
+      // when
+      modeling.moveElements(elements, { x: 0, y: 300 }, root);
```
Comment (barmac) [inline test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js]:
```suggestion
      modeling.moveElements(movedElements, { x: 0, y: 300 }, root);
```

File: test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js [hunk 5]
```diff
@@ -61,6 +60,58 @@ describe('features/modeling/behavior - detach events', function() {
...
+
+
+    it('should execute on batch', inject(function(canvas, elementRegistry, modeling) {
+
+      // given
+      var eventIds = [ 'BoundaryEventWithLabel', 'BoundaryEvent', 'BoundaryTimerEvent', 'BoundaryNonInterruptingMessageEvent', 'BoundaryEventWithConnections' ],
+          elements = eventIds.map(function(eventId) {
+            return elementRegistry.get(eventId);
+          }),
+          clone = elements.slice(0),
```
Comment (barmac) [inline test/spec/features/modeling/behavior/DetachEventBehaviorSpec.js]:
```suggestion
          clone = elements.slice(),
```
0 is default here

File: lib/features/modeling/behavior/DetachEventBehavior.js [hunk 6]
```diff
@@ -23,38 +23,16 @@ export default function DetachEventBehavior(eventBus, bpmnReplace) {
...
-        };
-      }
-
-      newShape = bpmnReplace.replaceElement(shape, intermediateEvent, { layoutConnection: false });
-
-      context.shapes = [ newShape ];
-    }
+    var shapes = context.shapes;
+
+    shapes.filter(function(shape) {
```
Comment (barmac) [inline lib/features/modeling/behavior/DetachEventBehavior.js]: Let's simplify this, ok? We can have a look it together tomorrow.

**Comment by oguzeroglu** [comment]
@barmac Thanks! Finished the refactoring related to your comments.

**Comment by philippfromme** [comment]
This has not been properly fixed guys. First of all you [monkey patch](https://github.com/bpmn-io/bpmn-js/pull/1206/files#diff-1ed47ece9adac0c5678433b1e2713d29R260) in BpmnReplace. Second, [AttachSupport](https://github.com/bpmn-io/diagram-js/blob/develop/lib/features/attach-support/AttachSupport.js#L118) should update the attachment of the shape. We should replace the shape AFTER that.

**Comment by oguzeroglu** [comment]
@philippfromme Should we implement another class like DetachSupport for this? If I remember correctly we once had this discussion that we don't wanna be able to attach multiple elements at once due to complications. If we get rid of [this line](https://github.com/bpmn-io/diagram-js/blob/2d8a56a21a70d2176f959a07c1569688b2be5424/lib/features/attach-support/AttachSupport.js#L126) in order to attach Batch detach which this PR addresses, then aren't we also allowing multiple shapes to be attach which would bring some cases we did not consider?

About the monkey patching stuff, [see this comment](https://github.com/bpmn-io/bpmn-js/commit/06a8a2c571c4fcf95befb98868a122e1636a8c65#r35428190).

If you could explain a bit more about what's on your mind about a proper solution, I can do a PR about this as well.

**Pull request merged**